### ISSUE TITLE
Registering lints with rustc

### ISSUE THREAD

**Issue by xFrednet** [rust-linting/design#14]
Rustc provides functionality to change the lint level in code, emit lints with rustc's diagnostics and notify the user if a lint was misspelled, renamed, or deprecated. These functions require the lints to be registered in rustc's [LintStore](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/struct.LintStore.html) and to be passed to lint emitting functions like [`struct_span_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/context/trait.LintContext.html#method.struct_span_lint). These functions take a [Lint](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/struct.Lint.html) reference with a static lifetime. This is problematic for the API, as we can't create a driver-specific type instance with a static lifetime without exposing the type to the user. 

This issue intends to investigate options to bridge this gap.

**Comment by xFrednet** [comment]
# 1. Unsafe lifetime extension.

The simplest solution is, probably, to create a driver specific `Lint` instance in our driver and just expand the lifetime to `'static`.  This would require unsafe code, to expand the lifetime requires us to figure out where we store the created lint instance, since we can't delete it afterwards to keep the lifetime as close to `'static` as we can.

## 1.1 Store the `driver::Lint` instance in the `api::Lint` instance

In the rust-linting prototype, the `api::Lint` instance is created using a macro and the `ast::Lint` type is not part of the stable API. Therefore, we could add an extra field to hold the driver lint instance. Every call that requires the `driver::Lint` instance would simply take the field from `ast::Lint`.

<details>
<summary>Mock/draft implementation</summary>

This is a mock implementation to showcase the general idea. This will not compile and would require unsafe code to convert the references between each other.

**API Code**

```rs
mod api {
    struct Lint {
        // ...
        /// The driver specific lint instance. This will be overridden by the driver
        driver_lint: OnceCell<&'static ()>,
    }
}
```

**Driver code**

```rs
// Like rustc's Lint
extern driver::Lint;

fn register(registered_lints: [&'static api::Lint]) {
    for api_lint in registered_lints {
        // This creates a driver lint instance and sets it's instance in `api::Lint`
        let driver_lint = Box<(diver::Lint::new(...));
        api_lint.driver_lint.set(driver_lint.leak() as &'static ());

        context.lint_store.register(api_lint.driver_lint as &'static driver::Lint)
    }
}

fn span_lint(api_lint: &'static api::Lint) {
    rust_context.span_lint(api_lint.driver_lint as &'static driver::Lint)
}
```

</details>

## 1.2 Use a Map

Instead of storing the `driver::Lint` instance in `api::Lint` we could create a map, stored in the `AstContext` to convert the types. The advantage of this is, that we would only need unsafe code once to extend the `driver::Lint` lifetime. Afterwards, everything will be safe. The disadvantage is, that each map lookup requires some performance.

<details>
<summary>Mock/draft implementation</summary>

This is a mock implementation to showcase the general idea. This will not compile and would require unsafe code to convert the references between each other.

**Driver code**

```rs
// Like rustc's Lint
extern driver::Lint;
extern api::Lint;

RustcContext {
    lints: HashMap<&'static api::Lint, &'static driver::Lint>,
}

impl RustcContext {

    fn register(&mut self, new_lint: &'static api::Lint) {
        let driver_lint_box = Box<(diver::Lint::new(...));
        let driver_lint = driver_lint_box.leak() as &'static driver::Lint;
        self.lints.put(new_lint, driver_lint);
        /// Getting `&'static driver::Lint` is now as simple as this
        // ```rs
        //     api_lint.lints[api_lint]
        // ```

        context.lint_store.register(driver_lint)
    }
    
    fn span_lint(self, api_lint: &'static api::Lint) {
        rust_context.span_lint(api_lint.lints[api_lint])
    }
}
```

</details>

# 2. Create `driver::Lint` instances at compile time.

As the issue says, we want to avoid exposing the `driver::Lint` type to the user. However, what we can do is change the `linter_adapter` implementation to not load static libraries, but instead create a single binary that includes all registered lint creates. This would enable us to inject some driver-specific code. As a disadvantage, it would mean that the adapter is no longer general and part of the driver implementation. It also feels like a break of the single responsibility principle. On the other hand, it would enable us to create `&'static driver::Lint` instances without unsafe hackery.

# 3. Change rustc's interface.

The main problem with rustc's `LintStore` interface, is that it takes a reference with a `'static` lifetime, while we would like to provide an instance with a limited lifetime. As a solution, we could try to change the interface of rustc. If rust-linting reaches a reasonable size, I can imagine that the diagnostics team would be fine with such a change, if it doesn't inconvenience the internal usage.

---

CC: @DevinR528, you've also created a prototype. Could you maybe explain how you did it. You mentioned something [here](https://github.com/rust-linting/design/issues/13#issuecomment-1125451413)

**Comment by DevinR528** [comment]
So the way I go about it is, in the driver (since that's the only place we can use `rustc` types) I have [conversion traits implemented](https://github.com/DevinR528/pluggy/blob/main/driver/src/register.rs#L205) for every type that makes up a [`stable::Lint`](https://github.com/DevinR528/pluggy/blob/main/pluggy_api/src/lint.rs) and after reading about the user provided `stable::Lint` (my `export_lint_pass` macro makes an ffi function that returns a `Vec<&'static stable::Lint>`, it also does the unsafe unspeced ffi `fn() -> *mut Box<dyn stable::LateLintImpl>`) we `Box::leak` each one and pass them to the `LintStore::register_lint` method [here](https://github.com/DevinR528/pluggy/blob/main/driver/src/register.rs#L761).

So, could we make the `stable::Lint` type more like rustc's `Lint` type and convert and leak?

**Comment by jhpratt** [comment]
Wait, I thought the goal was to have a standalone tool, not something that would hook into rustc.

**Comment by xFrednet** [comment]
Wasn't the goal to create a wrapper/stable API?

The way I designed rust-linting right now, is a tool that loads lint crates and provides them with a defined AST representation and later on diagnostic support. The backend, is exchangeable, currently I just use rustc, as I see little advantage in rewriting a lexer and parser.

All goals we've discussed are archivable with this architecture AFAIK. Or have I missed something. I actually wasn't aware that we wanted a completely standalone tool :thinking:

**Comment by jhpratt** [comment]
Yes, a wrapper/stable API were necessary. The lexer, parser, etc. could be reused from the `rustc_*` family of crates. I thought the actual emitting of diagnostics was going to be handled on our end, not rustc's.

But go ahead! You're the primary person doing anything on this front right now anyways. Don't let me stop you :slightly_smiling_face:

**Comment by xFrednet** [comment]
The diagnostic emission will also be exchangeable. I'll keep this in mind when I create the first version of the diagnostic emission. :upside_down_face:

**Comment by DevinR528** [comment]
If the driver crate depends on the `rustc_*` crates at all, is there a reason not to re-use as much as possible, at least where the pain of making it stable is worth the reuse from rustc?

**Comment by xFrednet** [comment]
> So, could we make the stable::Lint type more like rustc's Lint type and convert and leak?

We could, but that would bind us to the rustc implementation. Rustc also has some fields which are relevant to feature flags and tool lints, which we don't need or should store someone else. The current implementation was inspired by the lint definition, but still deviates. :)

> at least where the pain of making it stable is worth the reuse from rustc?

Would you mind explaining what you mean with this? @DevinR528

A sidenote, one "problem" of rustc is, that it's designed to be fast and for compilation. The incremental compilation model can sometimes make things more complicated. It definitely has its place in rustc, but for rust-linting I see more value in having a simpler interface.

**Comment by DevinR528** [comment]
I was trying to say that any reuse of rustc (like reusing the diagnostic/warning/error reporting) should be weighed against how hard/hacky it is to make a stable wrapper around said feature (so keeping with the diagnostic example, we need our `stable::Lint` type API stable and convertible to `rustc_lints::Lint` and all the extra hoops we have to jump through to make our types work with rustic).
